dbms/src/Storages/Page/V3/PageDirectory.h

```diff
@@ -82,48 +82,170 @@ class PageDirectorySnapshot : public DB::PageStorageSnapshot
 };
 using PageDirectorySnapshotPtr = std::shared_ptr<PageDirectorySnapshot>;
 
+// This is not a thread safe class
+// It must be used with outer synchronization
+class MultiVersionRefCount
+{
+public:
+    static MultiVersionRefCount * createFrom(const PageVersion & ver, Int64 ref_count)
+    {
+        if (ref_count == 1)
+        {
+            return nullptr;
+        }
+        return appendRefCount(nullptr, ver, ref_count);
+    }
+
+    static MultiVersionRefCount * copy(MultiVersionRefCount * container)
```

Comment:
why define all functions as static instead of using member functions?

Reply:
Because the object will just created when ref happend. In other cases it is just a nullptr.

Reply:
How about we make `MultiVersionRefCount` an object which contains `{}` of `std::unique_ptr<std::map<PageVersion, Int64>>` by default?  So that we don't need to use static method.

Reply:
BTW `std::unique_ptr` has no size overhead.

Reply:
Fixed.